---
title: "FeatureCounts explore"
author: "Emeline Favreau"
date: "15/02/2021"
output: html_document
---
Copyright 2021 Emeline Favreau, University College London.


#### Objective of analysis

Exploring the results of FeatureCounts per species.
We aim to have a high ratio of Assigned:Unassigned alignments.

#### Analysis steps:
- Obtaining data
- Aim 1: Megalopta genalis
- Aim 2: Polistes canadensis Patalano
- Aim 3: Polistes dominula Taylor
- Aim 4: Liostenogaster flavolineata Taylor



```{r load all the libraries, eval = TRUE, echo = FALSE, include = FALSE}
# get libraries
basic_libraries <- c("ggplot2",
                     "tidyverse")

for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```


```{r import Megalopta data, eval = TRUE, echo = FALSE, include = FALSE}
# https://www.pauloldham.net/importing-csv-files-into-r/
# list all files with the extension
# forward sequencing
megalopta_fwd <- list.files("Megalopta-genalis/result/Jones2017fwd/mapping",
                        pattern = "*1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                        full.names = TRUE) 

# reverse sequencing
megalopta_rev <- list.files("Megalopta-genalis/result/Jones2017rev/mapping",
                        pattern = "*1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                        full.names = TRUE) 

# combine
megalopta <- c(megalopta_fwd, megalopta_rev)

# create a list with datasets
megalopta_list <- lapply(megalopta,
                             read.delim)

# vector with names of samples of fwd seq
megalopta_fwd_names <- gsub(x = megalopta_fwd,
                             pattern = "Megalopta-genalis/result/Jones2017fwd/mapping/|_1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                             replacement = "")

# vector with names of samples of rev seq
megalopta_rev_names <- gsub(x = megalopta_rev,
                             pattern = "Megalopta-genalis/result/Jones2017rev/mapping/|_1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                             replacement = "")

# name each element as the sample name
names(megalopta_list) <- c(megalopta_fwd_names, megalopta_rev_names)

# import phenotype
megalopta.phenotype <- read.csv("Megalopta-genalis/megalopta-phenotype.csv",
                                header = FALSE,
                                stringsAsFactors = FALSE)


```

```{r import Polistes data, eval = TRUE, echo = FALSE, include = FALSE}
#https://www.pauloldham.net/importing-csv-files-into-r/
# list 10 files with the extension
polistes.canadensis <- list.files("Polistes-canadensis/result/Patalano/mapping",
                        pattern = "*1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                        full.names = TRUE) 

# create a list with datasets
polistes.canadensis_list <- lapply(polistes.canadensis,
                             read.delim)

# name each element as the sample name
names(polistes.canadensis_list) <- gsub(x = polistes.canadensis,
                                        pattern = "Polistes-canadensis/result/Patalano/mapping/|_1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                                        replacement = "")

# import phenotype
polistes.canadensis.phenotype <- read.csv("Polistes-canadensis/polistes-canadensis-patalano-phenotype.csv",
                                header = FALSE,
                                stringsAsFactors = FALSE)



```

```{r import Polistes dominula data, eval = TRUE, echo = FALSE, include = FALSE}
#https://www.pauloldham.net/importing-csv-files-into-r/
# list 10 files with the extension
polistes.dominula <- list.files("Polistes-dominula/result/Taylor/mapping",
                        pattern = "*1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                        full.names = TRUE) 

# create a list with datasets
polistes.dominula_list <- lapply(polistes.dominula,
                             read.delim)

# name each element as the sample name
names(polistes.dominula_list) <- gsub(x = polistes.dominula,
                                        pattern = "Polistes-dominula/result/Taylor/mapping/|_1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                                        replacement = "")

# import phenotype
polistes.dominula.phenotype <- read.csv("Polistes-dominula/polistes-dominula-taylor-phenotype.csv",
                                header = FALSE,
                                stringsAsFactors = FALSE)



```

```{r import Liostenogaster data, eval = TRUE, echo = FALSE, include = FALSE}
#https://www.pauloldham.net/importing-csv-files-into-r/
# list 10 files with the extension
liostenogaster.flavolineata <- list.files("Liostenogaster-flavolineata/result/Taylor/mapping",
                        pattern = "*1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                        full.names = TRUE) 

# create a list with datasets
liostenogaster.flavolineata_list <- lapply(liostenogaster.flavolineata,
                             read.delim)

# name each element as the sample name
names(liostenogaster.flavolineata_list) <- gsub(x = liostenogaster.flavolineata,
                                        pattern = "Liostenogaster-flavolineata/result/Taylor/mapping/|_1Aligned.sortedByCoord.out_gene.featureCounts.txt.summary",
                                        replacement = "")

# import phenotype
liostenogaster.flavolineata.phenotype <- read.csv("Liostenogaster-flavolineata/liostenogaster-flavolineata-taylor-phenotype.csv",
                                header = FALSE,
                                stringsAsFactors = FALSE)



```

```{r function to make a plot, eval = TRUE, echo = FALSE, include = TRUE}

# make a figure 
make_individual_plots <- function(pheno_table, sample_name){
  
  phenotype_table <- pheno_table
  
  this_run_name <- gsub(x = names(sample_name)[2],
                        pattern = "_1Aligned.sortedByCoord.out.bam",
                        replacement = "")
  
  this_pheno <- phenotype_table$phenotype[phenotype_table$samples ==
                                                           this_run_name]
  
  this_individual <- phenotype_table$individual_name[phenotype_table$samples ==
                                                           this_run_name]
  
 p <- ggplot(data = sample_name,
       aes(x = Status,
           y = sample_name[, 2])) +
    
  geom_bar(stat = "identity") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
    labs(title = paste(this_run_name, this_pheno, this_individual, sep = " ")) +
    ylab("Alignment Counts")
  
  return(p)
  #print(p)
}


```


### Megalopta genalis
We ran Nextflow pipeline twice, because samples were prepared using a strand aware library, resulting in half the samples presenting a forward strandedness status and the other half a reverse strandedness. The forward-stranded samples were run with Neftlow flag --forwardstranded, and the reverse-stranded samples --reverseStranded. This produced comparable results from featureCounts.

```{r Megalopta, eval = TRUE, echo = FALSE, include = TRUE}
# name col of phenotypes
colnames(megalopta.phenotype) <- c("samples",
                                   "individual_name",
                                   "phenotype")

# plot individual samples
# for (this_sample in 1:length(megalopta_list)){
#   print(make_individual_plots(megalopta.phenotype, megalopta_list[[this_sample]]))
# }

# make a violin plot for all
# combine info in a wide format dataset
megalopta_df_wd <- megalopta_list %>%
  reduce(left_join,
         by = "Status")

# change to long format
# three columns: count, sample, Status
megalopta_df_lg <- gather(megalopta_df_wd,
       sample,
       count,
       colnames(megalopta_df_wd)[2:length(colnames(megalopta_df_wd))])


# plot one violin per feature status, colour + jitter to help visualising
ggplot(megalopta_df_lg, aes(x    = Status,
                            y    = count,
                            colour = Status)) + 
  geom_violin() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1, size = 8)) +
  geom_jitter(aes(x = Status,
                  y = count,
                  colour = Status),
              shape = 16,
              position = position_jitter(0.2)) +
  labs(title = paste("Megalopta", "Jones 2017", "30 samples", sep = ", ")) +
  theme(legend.position = "none") +
  ylab("Alignment Counts") +
  xlab("Alignment status")

```



### Polistes canadensis

```{r Polistes canndensis, eval = TRUE, echo = FALSE, include = TRUE}
# name col of phenotypes
colnames(polistes.canadensis.phenotype) <- c("samples",
                                   "individual_name",
                                   "phenotype")
# 
# for (this_sample in 1:length(polistes.canadensis_list)){
#   print(make_individual_plots(polistes.canadensis.phenotype,
#                   polistes.canadensis_list[[this_sample]]))
# }
# make a violin plot for all
# combine info in a wide format dataset
polistes.canadensis_df_wd <- polistes.canadensis_list %>%
  reduce(left_join,
         by = "Status")

# change to long format
# three columns: count, sample, Status
polistes.canadensis_df_lg <- gather(polistes.canadensis_df_wd,
       sample,
       count,
       colnames(polistes.canadensis_df_wd)[2:length(colnames(polistes.canadensis_df_wd))])


# plot one violin per feature status, colour + jitter to help visualising
ggplot(polistes.canadensis_df_lg, aes(x    = Status,
                            y    = count,
                            colour = Status)) + 
  geom_violin() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1, size = 8)) +
  geom_jitter(aes(x = Status,
                  y = count,
                  colour = Status),
              shape = 16,
              position = position_jitter(0.2)) +
  labs(title = paste("P. canadensis", "Patalano 2015", "10 samples", sep = ", ")) +
  theme(legend.position = "none") +
  ylab("Alignment Counts") +
  xlab("Alignment status")

```

### Polistes dominula

```{r Polistes dominula, eval = TRUE, echo = FALSE, include = TRUE}
# name col of phenotypes
colnames(polistes.dominula.phenotype) <- c("samples",
                                   "individual_name",
                                   "phenotype")
# 
# for (this_sample in 1:length(polistes.dominula_list)){
#   print(make_individual_plots(polistes.dominula.phenotype,
#                   polistes.dominula_list[[this_sample]]))
# }
# make a violin plot for all
# combine info in a wide format dataset
polistes.dominula_df_wd <- polistes.dominula_list %>%
  reduce(left_join,
         by = "Status")

# change to long format
# three columns: count, sample, Status
polistes.dominula_df_lg <- gather(polistes.dominula_df_wd,
       sample,
       count,
       colnames(polistes.dominula_df_wd)[2:length(colnames(polistes.dominula_df_wd))])


# plot one violin per feature status, colour + jitter to help visualising
ggplot(polistes.dominula_df_lg, aes(x    = Status,
                            y    = count,
                            colour = Status)) + 
  geom_violin() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1, size = 8)) +
  geom_jitter(aes(x = Status,
                  y = count,
                  colour = Status),
              shape = 16,
              position = position_jitter(0.2)) +
  labs(title = paste("P. dominula", "Taylor 2020", "24 samples", sep = ", ")) +
  theme(legend.position = "none") +
  ylab("Alignment Counts") +
  xlab("Alignment status")

```

### Liostenogaster flavolineata

```{r Liostenogaster flavolineata, eval = TRUE, echo = FALSE, include = TRUE}
# name col of phenotypes
colnames(liostenogaster.flavolineata.phenotype) <- c("samples",
                                   "individual_name",
                                   "phenotype")
# 
# for (this_sample in 1:length(liostenogaster.flavolineata_list)){
#   print(make_individual_plots(liostenogaster.flavolineata.phenotype,
#                   liostenogaster.flavolineata_list[[this_sample]]))
# }
# make a violin plot for all
# combine info in a wide format dataset
liostenogaster.flavolineata_df_wd <- liostenogaster.flavolineata_list %>%
  reduce(left_join,
         by = "Status")

# change to long format
# three columns: count, sample, Status
liostenogaster.flavolineata_df_lg <- gather(liostenogaster.flavolineata_df_wd,
       sample,
       count,
       colnames(liostenogaster.flavolineata_df_wd)[2:length(colnames(liostenogaster.flavolineata_df_wd))])


# plot one violin per feature status, colour + jitter to help visualising
ggplot(liostenogaster.flavolineata_df_lg, aes(x    = Status,
                            y    = count,
                            colour = Status)) + 
  geom_violin() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1, size = 8)) +
  geom_jitter(aes(x = Status,
                  y = count,
                  colour = Status),
              shape = 16,
              position = position_jitter(0.2)) +
  labs(title = paste("L.flavolineata", "Taylor MS", "19 samples", sep = ", ")) +
  theme(legend.position = "none") +
  ylab("Alignment Counts") +
  xlab("Alignment status")

```

### Conclusion
So far, most of the aligned reads are assigned to features (such as exon, gene, CDS).




```{r record versions of session, eval = TRUE, echo = FALSE, include = FALSE}
# record versions of R and packages here
sessionInfo()
# R version 3.6.3 (2020-02-29)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Catalina 10.15.4
# 
# Matrix products: default
# BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] forcats_0.5.0   stringr_1.4.0   dplyr_0.8.5     purrr_0.3.4     readr_1.3.1    
# [6] tidyr_1.0.3     tibble_3.0.1    tidyverse_1.3.0 ggplot2_3.3.0  
# loaded via a namespace (and not attached):
#  [1] Rcpp_1.0.4.6     cellranger_1.1.0 pillar_1.4.4     compiler_3.6.3   dbplyr_1.4.3    
#  [6] tools_3.6.3      lubridate_1.7.8  jsonlite_1.6.1   lifecycle_0.2.0  gtable_0.3.0    
# [11] nlme_3.1-147     lattice_0.20-41  pkgconfig_2.0.3  rlang_0.4.6      reprex_0.3.0    
# [16] cli_2.0.2        DBI_1.1.0        rstudioapi_0.11  yaml_2.2.1       haven_2.2.0     
# [21] xfun_0.13        xml2_1.3.2       withr_2.2.0      httr_1.4.1       knitr_1.28      
# [26] fs_1.4.1         generics_0.0.2   vctrs_0.3.0      hms_0.5.3        grid_3.6.3      
# [31] tidyselect_1.1.0 glue_1.4.1       R6_2.4.1         fansi_0.4.1      readxl_1.3.1    
# [36] modelr_0.1.7     magrittr_1.5     scales_1.1.1     backports_1.1.7  ellipsis_0.3.0  
# [41] rvest_0.3.5      assertthat_0.2.1 colorspace_1.4-1 stringi_1.4.6    munsell_0.5.0   
# [46] broom_0.5.6      crayon_1.3.4 
```
