---
title: "Comparative analysis of SVM core predictor genes against DEG & WGCNA"
author: "Emeline Favreau"
date: "2021/12/13"
output: html_document
---

Copyright 2021 Emeline Favreau, University College London.

# Objective of analysis

We obtained 127 predictor genes from SVM models: these genes are common to all six species ( *Ceratina australensis*, *C. calcarata*, *Megalopta genalis*, *Polistes canadensis*, *P. dominula*, *Liostenogaster flavolineata*). We want to qualify them: are they also differentially expressed between reproductives and non-reproductives? are they also significantly associated with reproductive status in WGCNA?

## Analysis steps:
<<<<<<< HEAD

-   Obtaining data
-   Aim 1: Overlap of orthogroups (SVM, DEGs, WGCNA)
-   Aim 2: Common GO Terms enriched in all three analyses (SVM, DEGs, WGCNA)
=======
- Obtaining data
- Aim 1: Overlap of orthogroups (SVM, DEGs, WGCNA)
- Aim 2: Common GO Terms enriched in all three analyses (SVM, DEGs, WGCNA)
- Aim 3: Plot REVIGO
>>>>>>> f15ad1daf66748e80dc2552521f460ae8a6f8c93

```{r load all the libraries, eval = TRUE, echo = FALSE, include = FALSE}
# get libraries
basic_libraries <- c("UpSetR",
                     "tidyverse", 
                     "scales",
                     "ggplot2")

for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```

```{r import data, eval = TRUE, echo = FALSE, include = FALSE}
# common predictor genes (SVM)
predictor_genes <- 
  read.csv("../04-SVM/result/species-normalised/regression-on-all/127_common_predictor_genes",
  sep = "", stringsAsFactors = FALSE)

# common caste-specific genes (DEG)
de_genes <- 
  read.csv("../dge-DESeq2/orthology-dependent/orthogroups3718_all6spp_together/all_DEGs.txt",
  sep = "", stringsAsFactors = FALSE, header = FALSE)

# common co-expressed genes (WGCNA)
wgcna_genes <- 
  read.csv("../WGCNA/orthology-dependent/consensusNetwork_orthogroups3718/minModuleSize10/orthogroupList_SigModule-SigTraitAssociated.txt",
  sep = "", stringsAsFactors = FALSE, header = FALSE)


##################################
# GO terms for 127 predictor genes (all categories)
predictor_go <- 
  read.csv("../svm/result/species-normalised/regression-on-all/topgo_result_all_species_all_categories",
  sep = "\t", stringsAsFactors = FALSE)

# GO terms for common caste-specific genes (DEG, BP only)
de_go <- 
  read.csv("../ortho-enrichment/resultOnScratch/topgo_result_consensus-degs_BP",
  sep = "\t", stringsAsFactors = FALSE)

# GO terms common co-expressed genes (WGCNA)
wgcna_go <- 
  read.csv("../WGCNA/result/module_size_10/all/topgo_result_BP",
  sep = "\t", stringsAsFactors = FALSE)


```

# Aim 1: Overlap of orthogroups (SVM, DEGs, WGCNA)

```{r aim 1 orthogroup overlap, eval = TRUE, echo = FALSE, include = TRUE}
# comparative analysis of orthogroups 
# common to all three
common_svm_degs_vec <- predictor_genes$predictor_gene[predictor_genes$predictor_gene %in% de_genes$V1]
common_svm_wgcna_vec <- predictor_genes$predictor_gene[predictor_genes$predictor_gene %in% wgcna_genes$V1]
common_svm_degs_vec[common_svm_degs_vec %in% wgcna_genes$V1]


# make a vector for each species with the OG-orthogroups that are DEGs and SVMs
OG_listInput <- list(
  predictor_genes = predictor_genes$predictor_gene,
                           
  de_genes = de_genes$V1,
                           
  wgcna_genes = wgcna_genes$V1)

# str(OG_listInput)

upset(fromList(OG_listInput),
      sets = c("predictor_genes",
                "de_genes",
                "wgcna_genes"),
      order.by = "degree", keep.order = TRUE)

# keep a copy as pdf
pdf("DEGs_SVM_WGCNA_comparative_plot.pdf")
# plot comparative analysis of species DEGs in SVM core genes
upset(fromList(OG_listInput),
      sets = c("predictor_genes",
                "de_genes",
                "wgcna_genes"),
      order.by = "degree", keep.order = TRUE)# 
dev.off()
```

There is little overlap between the three genes sets. Two genes are common to all three: "OG0000706" (uncharacterized protein LOC408365 isoform X2 [Apis mellifera]) and "OG0001935" (uncharacterized protein LOC408661 isoform X1 [Apis mellifera]).

# Aim 2: Common GO Terms enriched in all three analyses (SVM, DEGs, WGCNA)

```{r aim 2 GO terms, eval = TRUE, echo = FALSE, include = TRUE}
# comparative analysis of GO Terms (BP)
predictor_go_BP_vec <- predictor_go$GO.ID[predictor_go$goCategory == "BP"]
de_go_BP_vec <- de_go$GO.ID[de_go$goCategory == "BP"]
wgcna_go_BP_vec <- wgcna_go$GO.ID[wgcna_go$goCategory == "BP"]

# common to all three: 15396
commongo_svm_degs_vec <- predictor_go_BP_vec[predictor_go_BP_vec %in% de_go_BP_vec]
commongo_svm_wgcna_vec <- predictor_go_BP_vec[predictor_go_BP_vec %in% wgcna_go_BP_vec]
commongo_svm_wgcna_deg_vec <- commongo_svm_degs_vec[commongo_svm_degs_vec %in% wgcna_go_BP_vec]

# vector of all possible GO terms (4226)
all_go_vec <- unique(c(predictor_go_BP_vec, de_go_BP_vec, wgcna_go_BP_vec))

# vector of Go terms common to all (2657)
common_go_vec <- c()

for(this_go in all_go_vec){
  if(this_go %in% predictor_go_BP_vec){
    if(this_go %in% de_go_BP_vec){
      if(this_go %in% wgcna_go_BP_vec){
  
        common_go_vec <- c(common_go_vec, this_go)
      }
    }
  }
}


# make a vector for each category
GO_listInput <- list(
  predictor_GO = predictor_go_BP_vec,
                           
  de_GO = de_go_BP_vec,
                           
  wgcna_GO = wgcna_go_BP_vec)

# str(GO_listInput)

upset(fromList(GO_listInput),
      sets = c("predictor_GO",
                "de_GO",
                "wgcna_GO"),
      order.by = "degree", keep.order = TRUE)

# keep a copy as pdf
pdf("DEGs_SVM_WGCNA_GO_comparative_plot.pdf")
# plot comparative analysis of species DEGs in SVM core genes
upset(fromList(GO_listInput),
      sets = c("predictor_GO",
                "de_GO",
                "wgcna_GO"),
      order.by = "degree", keep.order = TRUE)
dev.off()




#######
# REVIGO will reduce the GO terms for us, prvided we give a list of 2000 GO IDs
# Here we select the 2000 GO terms that have the lowest cumulative pvalue from the three analyses

# selecting the threshold for including GO terms into REVIGO analysis
svm_mean <- mean(predictor_go$pvalue[predictor_go$goCategory == "BP"])
deg_mean <- mean(de_go$pvalue[de_go$goCategory == "BP"])
wgcna_mean <- mean(wgcna_go$pvalue[wgcna_go$goCategory == "BP"])

# lowest cumulative p-value
common_go_pvalue_df <- data.frame(go = common_go_vec,
                                  svm_pvalue = NA,
                                  deg_pvalue = NA,
                                  wgcna_pvalue = NA,
                                  cumulative_pvalue = NA)


for(i in 1:length(common_go_vec)){
  # select pvalue from svm (mean of the six analyses)
  this_svm_mean <- mean(predictor_go$pvalue[predictor_go$goCategory == "BP" &
                     predictor_go$GO.ID == common_go_vec[i]])
  
  # select pvalue from DEGs (consensus, only one value)
  this_deg_mean <- mean(de_go$pvalue[de_go$goCategory == "BP" &
                      de_go$GO.ID == common_go_vec[i]])
  
  # select pvalue from WGCNAs (consensus, only one value)
  this_wgcna_mean <- mean(wgcna_go$pvalue[wgcna_go$goCategory == "BP" &
                        wgcna_go$GO.ID == common_go_vec[i]])
  
  this_cumulative_pvalue <- sum(this_svm_mean, this_wgcna_mean, this_deg_mean)
  # add result to table
  common_go_pvalue_df[i, 2:5] <- c(this_svm_mean, this_deg_mean,
                               this_wgcna_mean, this_cumulative_pvalue)
}

# check result
summary(common_go_pvalue_df$cumulative_pvalue)

# select the 2000 GO terms (max needed by REVIGO) whose cumulative pvalue is lowest
common_go_pvalue_ordered <- 
  common_go_pvalue_df[order(common_go_pvalue_df$cumulative_pvalue), ]

# select 2000 tops 
common_go_for_revigo <- common_go_pvalue_ordered$go[1:2000]

# save list of GO terms, as an input to REVIGO
write.table(common_go_for_revigo,
            file = "common_GO_terms_svm_deg_wgcna",
            row.names = FALSE,
            quote = FALSE,
            col.names = FALSE)


```

# Aim 3: Plot REVIGO result

```{r prepare REVIGO, eval = TRUE, echo = FALSE, include = TRUE}
# A plotting R script produced by the Revigo server at http://revigo.irb.hr/
# If you found Revigo useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# --------------------------------------------------------------------------
# If you don't have the ggplot2 package installed, uncomment the following line:
# install.packages( "ggplot2" );
library( ggplot2 );

# --------------------------------------------------------------------------
# If you don't have the scales package installed, uncomment the following line:
# install.packages( "scales" );
library( scales );

# for colour gradient
library(RColorBrewer)

# --------------------------------------------------------------------------
# Here is your data from Revigo. Scroll down for plot configuration options.

revigo.names <- c("term_ID","description","frequency","plot_X","plot_Y","log_size","uniqueness","dispensability");
revigo.data <- rbind(c("GO:0008150","biological_process",100,3.43937175384303,8.18836680266251,4.06088662300466,1,0),
c("GO:0022610","biological adhesion",1.7298331015299,1.62168660916902,8.70175946996726,2.30102999566398,1,0),
c("GO:0030029","actin filament-based process",2.38178025034771,2.70615465050388,7.97658586272885,2.43933269383026,0.968538916387674,0),
c("GO:0030155","regulation of cell adhesion",0.460709318497914,0.0675393687774227,-3.15641813539858,1.73239375982297,0.939069219436687,0),
c("GO:0032501","multicellular organismal process",30.7892906815021,2.33779219837252,8.52776086754993,3.54937115233318,1,0),
c("GO:0032502","developmental process",24.6696801112656,-3.08436735807402,-1.62455353723656,3.45316539252586,1,0),
c("GO:0040007","growth",1.53859527121001,-4.47087893571243,-5.28310717647774,2.25042000230889,1,0),
c("GO:0042908","xenobiotic transport",0.0608484005563282,-0.668010386629098,-0.31412743520023,0.903089986991944,0.997843426354159,0),
c("GO:0046034","ATP metabolic process",1.10396383866481,-7.14966937461513,0.750790709820956,2.10720996964787,0.91576597125256,0),
c("GO:0061061","muscle structure development",1.76460361613352,6.55891653579193,1.75904781043659,2.3096301674259,0.87510740444768,0),
c("GO:0051261","protein depolymerization",0.13038942976356,-5.3266513308534,-3.52122732314929,1.20411998265592,0.951077713098245,0.02386156),
c("GO:0007155","cell adhesion",1.66029207232267,7.63724919887819,-2.41913595236735,2.28330122870355,0.946471819804447,0.03160952),
c("GO:0042866","pyruvate biosynthetic process",0.00869262865090403,-1.41958401230013,7.35115673974767,0.301029995663981,0.792829852539171,0.07380535),
c("GO:0031647","regulation of protein stability",0.504172461752434,5.41009171721616,-3.61369414322919,1.77085201164214,0.925156782487924,0.09662859),
c("GO:0140053","mitochondrial gene expression",1.0778859527121,-6.09582673434277,1.53513376086855,2.09691001300806,0.900961790885003,0.11317515),
c("GO:0006091","generation of precursor metabolites and energy",1.76460361613352,-5.94783360069602,3.51023953965509,2.3096301674259,0.877914406534965,0.11970488),
c("GO:0022407","regulation of cell-cell adhesion",0.278164116828929,-2.62188736974821,-5.24595984490932,1.51851393987789,0.927812814239829,0.12366093),
c("GO:0043244","regulation of protein-containing complex disassembly",0.339012517385257,-0.78909946864899,-5.97952742443353,1.60205999132796,0.875353930815316,0.1258714),
c("GO:0046939","nucleotide phosphorylation",0.295549374130737,-5.02304845835101,5.59244810454441,1.54406804435028,0.758374891878358,0.13496975),
c("GO:0040008","regulation of growth",2.70340751043115,2.37386916313893,-4.41669652127471,2.49415459401844,0.932546361970041,0.13847828),
c("GO:0005975","carbohydrate metabolic process",2.21662030598053,-4.07615865772206,1.69722030367297,2.40823996531185,0.894058982183194,0.13851428),
c("GO:0035383","thioester metabolic process",0.49547983310153,-6.68761524249984,2.53630004540781,1.76342799356294,0.864626936072359,0.14243259),
c("GO:0043068","positive regulation of programmed cell death",0.634561891515994,-0.662903765367544,-3.92143269800228,1.86923171973098,0.917331070508953,0.14359166),
c("GO:0022900","electron transport chain",0.817107093184979,-4.77834815023347,2.14261440955093,1.97772360528885,0.827697985960447,0.15048865),
c("GO:0006298","mismatch repair",0.113004172461752,-1.36018433534697,5.02330403380659,1.14612803567824,0.813741900098453,0.15243188),
c("GO:0051960","regulation of nervous system development",1.82545201668985,1.73359445642954,-3.5414617102822,2.32428245529769,0.911569911857088,0.16351888),
c("GO:1901135","carbohydrate derivative metabolic process",4.44193324061196,-5.81135059652301,2.57034344462154,2.70926996097583,0.892155308251765,0.16684605),
c("GO:0044087","regulation of cellular component biogenesis",2.70340751043115,1.29184233399896,-5.0629378180965,2.49415459401844,0.926864916631047,0.17241081),
c("GO:0051049","regulation of transport",2.81641168289291,2.46186533210533,-5.53748144198777,2.51188336097887,0.912152503015101,0.1733941),
c("GO:0065008","regulation of biological quality",11.4568845618915,3.09623089680289,-4.41245892663075,3.12024479554637,0.930496335474482,0.17974051),
c("GO:0048865","stem cell fate commitment",0.0521557719054242,4.87283691310075,5.07586103549191,0.845098040014257,0.862474251748109,0.18184296),
c("GO:0007528","neuromuscular junction development",0.712795549374131,-5.16452655392498,-3.01277225234449,1.91907809237607,0.942800053957288,0.19119773),
c("GO:0051128","regulation of cellular component organization",6.00660639777469,1.55700530953245,-4.67841785338862,2.84010609445676,0.919659561568044,0.21688178),
c("GO:0033227","dsRNA transport",0.217315716272601,0.0246716729998148,-0.0679813405462356,1.41497334797082,0.997843426354159,0.2211681),
c("GO:0007030","Golgi organization",0.686717663421419,-6.22702397995916,-2.13996051752516,1.90308998699194,0.939307020712592,0.22320418),
c("GO:0007422","peripheral nervous system development",0.478094575799722,6.17342360552816,3.71453233551333,1.7481880270062,0.847029829314483,0.23641904),
c("GO:0060560","developmental growth involved in morphogenesis",0.678025034770515,6.22538592096456,1.6868109043972,1.89762709129044,0.863693527359708,0.24556523),
c("GO:0097435","supramolecular fiber organization",1.79937413073713,-5.77220478841253,-2.94961638366271,2.31806333496276,0.936998368802527,0.24719648),
c("GO:0044238","primary metabolic process",36.500347705146,-4.97745644392508,3.13640439967262,3.6232492903979,0.869202176440184,0.24811188),
c("GO:0072525","pyridine-containing compound biosynthetic process",0.104311543810848,-1.34581319635707,5.8920413787152,1.11394335230684,0.754999128842232,0.25746114),
c("GO:0120031","plasma membrane bounded cell projection assembly",1.42559109874826,-5.60463230214546,-2.20372905456834,2.21748394421391,0.928135183149067,0.26791939),
c("GO:0006508","proteolysis",7.80598052851182,-3.80472989998485,3.24464164136834,2.95375969173323,0.846505396140612,0.28742626),
c("GO:0048736","appendage development",2.84248956884562,5.13805416674905,2.34640688115636,2.51587384371168,0.850901459625776,0.29188913),
c("GO:0043603","cellular amide metabolic process",4.45931849791377,-3.27141228151021,4.37636642050973,2.71096311899528,0.769353198016491,0.30676233),
c("GO:0007517","muscle organ development",0.721488178025035,6.38588619765316,3.14405554782789,1.92427928606188,0.83657320479145,0.31809974),
c("GO:0048872","homeostasis of number of cells",0.312934631432545,5.71201586522077,-3.58251729810155,1.56820172406699,0.924442066654793,0.34699604),
c("GO:0050821","protein stabilization",0.347705146036161,5.56905042390409,-3.25669787860835,1.61278385671974,0.927660098479696,0.35033434),
c("GO:0030036","actin cytoskeleton organization",2.23400556328234,-6.0152409550043,-2.47273329591868,2.41161970596323,0.926796449933585,0.36271294),
c("GO:0010160","formation of animal organ boundary",0.312934631432545,5.07136037918279,3.22492206693987,1.56820172406699,0.825376162304214,0.38491997),
c("GO:0010001","glial cell differentiation",0.365090403337969,5.57023944326946,3.98930145894354,1.63346845557959,0.823094078079497,0.39594807),
c("GO:0009653","anatomical structure morphogenesis",13.0302503477051,5.15754798088195,2.7302301844941,3.17609125905568,0.835360641511479,0.39613127),
c("GO:0016043","cellular component organization",23.5048678720445,-5.61978945953199,-2.63927141587679,3.43216726944259,0.915993041220351,0.39642481),
c("GO:0048598","embryonic morphogenesis",2.03407510431154,5.85485250568632,2.164448129622,2.37106786227174,0.832312654863593,0.39878039),
c("GO:0000045","autophagosome assembly",0.269471488178025,-5.42162788677754,-0.522831058384744,1.50514997831991,0.849145426409093,0.41045575),
c("GO:0090066","regulation of anatomical structure size",2.07753824756606,4.67380943741687,-3.76656538737295,2.38021124171161,0.913880653883805,0.41867268),
c("GO:0072521","purine-containing compound metabolic process",1.79068150208623,-2.92715912394134,4.33523900918662,2.31597034545692,0.767953952031229,0.42558916),
c("GO:0014016","neuroblast differentiation",0.425938803894298,5.76719824958138,3.98850737181047,1.69897000433602,0.819420875985383,0.43137807),
c("GO:0035152","regulation of tube architecture, open tracheal system",0.791029207232267,5.35111202503402,0.332359110541821,1.96378782734556,0.781361181853839,0.44031443),
c("GO:1901879","regulation of protein depolymerization",0.252086230876217,-1.12097650533298,-5.9297929809993,1.47712125471966,0.869480900788654,0.45302077),
c("GO:1902904","negative regulation of supramolecular fiber organization",0.295549374130737,-1.06502146471021,-5.74485209979435,1.54406804435028,0.86326550693735,0.45928084),
c("GO:0045466","R7 cell differentiation",0.425938803894298,5.89507588419224,3.52438004418627,1.69897000433602,0.785964078043602,0.46337958),
c("GO:0032535","regulation of cellular component size",1.53859527121001,-1.80344229705436,-2.82475528786166,2.25042000230889,0.855026425602251,0.47671185),
c("GO:0030030","cell projection organization",5.56328233657858,-5.42767936378714,-2.70284819228383,2.80685802951882,0.928175324246944,0.47982336),
c("GO:0032787","monocarboxylic acid metabolic process",1.64290681502086,-3.01793393888087,6.53893480271344,2.27875360095283,0.714813414213082,0.48917827),
c("GO:0043604","amide biosynthetic process",3.19019471488178,-2.66401099874479,4.78353490450908,2.56584781867352,0.72242408544445,0.49166739),
c("GO:0034404","nucleobase-containing small molecule biosynthetic process",0.113004172461752,-2.00846156788483,6.27076419256753,1.14612803567824,0.681234307080124,0.49551413),
c("GO:0060562","epithelial tube morphogenesis",3.93776077885953,6.05804427021328,2.46041200234889,2.6570558528571,0.805845180204976,0.4971757),
c("GO:0001558","regulation of cell growth",0.747566063977747,-0.424144966585973,-5.78263605125672,1.93951925261862,0.863302798220558,0.49955363),
c("GO:0007399","nervous system development",9.17941585535466,5.81769640962596,3.05249789144857,3.02407498730743,0.799501221182104,0.50104931),
c("GO:0120032","regulation of plasma membrane bounded cell projection assembly",0.425938803894298,-0.763048522846637,-5.74779732471572,1.69897000433602,0.847836520390356,0.51056574),
c("GO:0019637","organophosphate metabolic process",3.72044506258693,-4.56996648741733,4.68624582189406,2.63245729218472,0.699414917150503,0.51918262),
c("GO:0046716","muscle cell cellular homeostasis",0.286856745479833,5.78947550777773,-3.43341689891045,1.53147891704226,0.90144751545671,0.52390534),
c("GO:0060581","cell fate commitment involved in pattern specification",0.304242002781641,4.90188683858793,4.15861969863429,1.55630250076729,0.827142993159921,0.52927199),
c("GO:0042592","homeostatic process",4.17246175243394,4.3159110857534,-3.85335447815869,2.68214507637383,0.907102036315982,0.5441645),
c("GO:0046434","organophosphate catabolic process",0.45201668984701,-4.9883976142414,4.96476691907052,1.72427586960079,0.690777878857086,0.54456564),
c("GO:0009166","nucleotide catabolic process",0.165159944367177,-3.39860519174419,5.60473395179733,1.30102999566398,0.648403389231574,0.54847917),
c("GO:1902903","regulation of supramolecular fiber organization",1.04311543810848,-0.34748258713082,-5.5664549366141,2.08278537031645,0.861950351859316,0.55891083),
c("GO:0031344","regulation of cell projection organization",1.26912378303199,-0.107587252286435,-5.6736053730272,2.16731733474818,0.859324002612106,0.59203685),
c("GO:0007400","neuroblast fate determination",0.234700973574409,5.30888260691477,4.19029234462977,1.44715803134222,0.80528418864645,0.59645217),
c("GO:0045664","regulation of neuron differentiation",0.695410292072323,1.16002122413515,-3.51032751274818,1.90848501887865,0.911483077839117,0.5983692),
c("GO:0060249","anatomical structure homeostasis",1.22566063977747,5.03104878206004,-3.68543386582695,2.15228834438306,0.914320021610464,0.59872795),
c("GO:0009167","purine ribonucleoside monophosphate metabolic process",0.208623087621697,-2.99812925196013,6.01408098101404,1.39794000867204,0.64463046007528,0.60818303),
c("GO:0007275","multicellular organism development",20.0712795549374,5.50325199293107,2.60018050644609,3.36361197989214,0.823251668535128,0.60862563),
c("GO:0051129","negative regulation of cellular component organization",1.34735744089013,-0.122056517996504,-5.52761102275684,2.19312459835446,0.854516790106857,0.60950564),
c("GO:0048666","neuron development",4.82440890125174,5.57876968751858,3.4543363636047,2.74507479158206,0.762312009151439,0.61634559),
c("GO:0007475","apposition of dorsal and ventral imaginal disc-derived wing surfaces",0.147774687065369,6.5984164897064,2.42094851145699,1.25527250510331,0.826947202865994,0.61697867),
c("GO:0006414","translational elongation",0.226008344923505,-1.95249948947782,4.61519459641426,1.43136376415899,0.776385098025022,0.62311922),
c("GO:0010004","gastrulation involving germ band extension",0.373783031988873,5.87254957307916,1.73691189726409,1.64345267648619,0.856870569159901,0.62660543),
c("GO:0048731","system development",14.7166203059805,5.60023057422872,2.81168955998735,3.22891340599469,0.81155893079291,0.62965645),
c("GO:0072522","purine-containing compound biosynthetic process",0.999652294853964,-2.26215560681539,5.05311602291485,2.06445798922692,0.691778719174314,0.62969189),
c("GO:0009887","animal organ morphogenesis",6.2413073713491,5.94548340981863,2.81026253579517,2.85672889038288,0.785350082464854,0.63314012),
c("GO:0055086","nucleobase-containing small molecule metabolic process",2.45132127955494,-2.70245927578833,5.40904110010598,2.45178643552429,0.691736326272488,0.63604134),
c("GO:0009141","nucleoside triphosphate metabolic process",0.356397774687065,-3.2999229978832,5.78434699743197,1.6232492903979,0.633988563294658,0.64779373),
c("GO:0030031","cell projection assembly",1.56467315716273,-5.75062740913094,-2.15414240893788,2.25767857486918,0.929753259275444,0.64896164),
c("GO:0034765","regulation of ion transmembrane transport",0.512865090403338,2.18916562726535,-6.07479932159636,1.77815125038364,0.918009421819143,0.65708743),
c("GO:0034762","regulation of transmembrane transport",0.521557719054242,2.31071091581507,-6.0275499090615,1.78532983501077,0.917884477861861,0.65897292),
c("GO:0009135","purine nucleoside diphosphate metabolic process",0.278164116828929,-3.14616054184562,5.91754327401745,1.51851393987789,0.62699831215685,0.66396846),
c("GO:0033866","nucleoside bisphosphate biosynthetic process",0.278164116828929,-2.86892448355822,5.75401637141678,1.51851393987789,0.61071767107447,0.66396846),
c("GO:0019693","ribose phosphate metabolic process",1.69506258692629,-4.63977586739217,4.98823374591788,2.29225607135648,0.657259718338856,0.66877485),
c("GO:0006090","pyruvate metabolic process",0.356397774687065,-2.45084140587855,7.10699920367314,1.6232492903979,0.749122813282161,0.67558057),
c("GO:0009132","nucleoside diphosphate metabolic process",0.356397774687065,-3.15012021890833,5.75095358909067,1.6232492903979,0.633988563294658,0.67856479),
c("GO:0009123","nucleoside monophosphate metabolic process",0.382475660639778,-3.24627065348758,5.64616606081947,1.65321251377534,0.631904739942509,0.68284203),
c("GO:0006122","mitochondrial electron transport, ubiquinol to cytochrome c",0.121696801112656,0.491362315437033,7.89580889752753,1.17609125905568,0.840108332442972,0.68862894),
c("GO:0001709","cell fate determination",1.15611961057024,5.06576951589989,4.37023709959539,2.12710479836481,0.821896573806191,0.68919049),
c("GO:0043269","regulation of ion transport",0.782336578581363,2.75874070603183,-5.92546134194591,1.95904139232109,0.919444079483756,0.69003973),
c("GO:0048856","anatomical structure development",23.4787899860918,5.35131540099931,2.7626821052199,3.43168534468601,0.826295110780561,0.69023619),
c("GO:0008587","imaginal disc-derived wing margin morphogenesis",0.45201668984701,6.39062996195844,2.59485188527249,1.72427586960079,0.80803497707832,0.69687439),
c("GO:0048513","animal organ development",10.1182197496523,5.67386028331334,3.06849240592411,3.06632592536204,0.796605732360699,0.69851688));

```

```{r plot save REVIGO, eval = TRUE, echo = FALSE, include = TRUE}
one.data <- data.frame(revigo.data);
names(one.data) <- revigo.names;
one.data <- one.data [(one.data$plot_X != "null" & one.data$plot_Y != "null"), ];
one.data$plot_X <- as.numeric( as.character(one.data$plot_X) );
one.data$plot_Y <- as.numeric( as.character(one.data$plot_Y) );
one.data$log_size <- as.numeric( as.character(one.data$log_size) );
one.data$frequency <- as.numeric( as.character(one.data$frequency) );
one.data$uniqueness <- as.numeric( as.character(one.data$uniqueness) );
one.data$dispensability <- as.numeric( as.character(one.data$dispensability) );
#head(one.data);


# --------------------------------------------------------------------------
# Names of the axes, sizes of the numbers and letters, names of the columns,
# etc. can be changed below

p1 <- ggplot( data = one.data );

p1 <- p1 + geom_point( aes( plot_X, plot_Y,
                            colour = uniqueness, size = log_size), 
                       alpha = I(0.6) ) + scale_size_area();

# make a divergent gradient
mydivcolvec <- brewer.pal(9, "GnBu")

# make the dot coloured based on uniqueness (darker=more unique)
p1 <- p1 + scale_colour_gradientn(colours = mydivcolvec,
                                   limits = c( min(one.data$uniqueness), 1) );

# add black around each dot
p1 <- p1 + geom_point( aes(plot_X, plot_Y, size = log_size), shape = 21, 
                       fill = "transparent", colour = I (alpha ("black", 0.6) )) 

# keep data for which go term is dispensable below 0.15
# dispensability: how much redundant is the term; kind of like uniqueness
ex <- one.data [ one.data$dispensability < 0.15, ];


# meaningful list
# including
# chromatin-associated: regulation of cellular component organization and
# "actin filament-based process"
# telomere associated: "nucleotide catabolic process"
# 
meaningful_list <-  c(  "actin filament-based process",
                        "cell fate determination",
                       "developmental process",
                       "embryonic morphogenesis",
                       "formation of animal organ boundary",
                       "growth",
                       "imaginal disc-derived wing margin morphogenesis",
                       #"negative regulation of supramolecular fiber organization",
                       #"neuromuscular junction development",
                       "nucleotide catabolic process",
                       #"positive regulation of programmed cell death",
                       "primary metabolic process",
                       #"regulation of anatomical structure size",
                       #"regulation of cell growth",
                       "regulation of cellular component biogenesis",
                       "regulation of cellular component organization",
                       #"regulation of growth",
                       #"regulation of nervous system development",
                       "stem cell fate commitment",
                       "supramolecular fiber organization")

# shorten the data to those meaningful terms
ex_meaningful <- one.data [one.data$description  %in% meaningful_list, ]


# label those GO Terms (choose betwween ex_meaningful or one.data)
p1 <- p1 + geom_text( data = ex_meaningful,
                      aes(plot_X, plot_Y, label = description),
                      colour = I(alpha("black", 0.85)), size = 3 );

# add axis labels
p1 <- p1 + labs (y = "semantic space y", x = "semantic space x");

# add a legend
p1 <- p1 + theme(legend.key = element_blank()) ;

# recenter the plot
one.x_range = max(one.data$plot_X) - min(one.data$plot_X);
one.y_range = max(one.data$plot_Y) - min(one.data$plot_Y);
p1 <- p1 + xlim(min(one.data$plot_X)-one.x_range/10,
                max(one.data$plot_X)+one.x_range/10);
p1 <- p1 + ylim(min(one.data$plot_Y)-one.y_range/10,
                max(one.data$plot_Y)+one.y_range/10);

# add a legend
p1 <- p1 + theme_bw() ;

# --------------------------------------------------------------------------
# Output the plot to screen

p1;

# Uncomment the line below to also save the plot to a file.
# The file type depends on the extension (default=pdf).

ggsave("2022-06-17-genetic-toolkit-revigo-plot.pdf");







```
