---
title: "DESeq2 Analyses: 3718 Orthogroups"
author: "KS Geist"
date: "17 July 2021"
output: html_document
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message = F, warning = F, include = F}
#install.packages("htmltools")
#BiocManager::install("DESeq2")
## Install BiocManager if you don't already have it first.

## Clear workspace
rm(list=ls())

## Set options
options(scipen=999)

## Load libraries
library(RCurl)
library(DESeq2)
library(htmltools)
library(vsn)
library(ggplot2)
library(tidyverse)
library(pheatmap)
```

## Getting Started

##### Read in the metadata:
```{r}
meta <- read.csv(file ="~/Dropbox/0.ISU/0.network_analyses/deg-DESeq2/Comparative_Transcriptomics_Metadata.csv", strip.white = T, header = T)
#str(meta)
```

##### Set the species:
```{r}
species <- c("Ceratina_australensis", "Ceratina_calcarata", "Megalopta_genalis", "Polistes_canadensis", "Polistes_dominula", "Liostenogaster_flavolineata") 
wasps <- species[4:6]
bees <- species[1:3]
```

##### Import the 3718 orthorgroups for filtering:
```{r}
Caus.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Ceratina_australensis_5787_gene_orthogroups_list", sep = "\t", header = F)
Ccal.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Ceratina_calcarata_5787_gene_orthogroups_list", sep = "\t", header = F)
beesOrthoData <- merge(Caus.ortho, Ccal.ortho, by = "V1")
colnames(beesOrthoData) <- c("Orthogroup", species[1], species[2])
# nrow(beesOrthoData)
# beesOrthoData
Mgen.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Megalopta_genalis_5787_gene_orthogroups_list", sep = "\t", header = F)
beesOrthoData <- merge(beesOrthoData, Mgen.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(beesOrthoData) <- c("Orthogroup", species[1], species[2], species[3])
# nrow(beesOrthoData)
# beesOrthoData

Pcan.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Polistes_canadensis_6983_gene_orthogroups_list", sep = "\t", header = F)
Pdom.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Polistes_dominula_6983_gene_orthogroups_list", sep = "\t", header = F)
waspsOrthoData <- merge(Pcan.ortho, Pdom.ortho, by = "V1")
colnames(waspsOrthoData) <- c("Orthogroup", species[4], species[5])
# nrow(waspsOrthoData)
# waspsOrthoData
Lfla.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Liostenogaster_flavolineata_6983_gene_orthogroups_list", sep = "\t", header = F)
waspsOrthoData <- merge(waspsOrthoData, Lfla.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(waspsOrthoData) <- c("Orthogroup", species[4], species[5], species[6])
# nrow(waspsOrthoData)
# waspsOrthoData

# head(beesOrthoData)
# head(waspsOrthoData)

## Also, read in and create a hymenoptera orthogroup df:
Caus.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Ceratina_australensis_3718_gene_orthogroups_list", sep = "\t", header = F)
Ccal.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Ceratina_calcarata_3718_gene_orthogroups_list", sep = "\t", header = F)
hymnOrthoData <- merge(Caus.ortho, Ccal.ortho, by = "V1")
colnames(hymnOrthoData) <- c("Orthogroup", species[1], species[2])
# nrow(hymnOrthoData)

Mgen.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Megalopta_genalis_3718_gene_orthogroups_list", sep = "\t", header = F)
hymnOrthoData <- merge(hymnOrthoData, Mgen.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(hymnOrthoData) <- c("Orthogroup", species[1], species[2], species[3])
# nrow(hymnOrthoData)

Pcan.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Polistes_canadensis_3718_gene_orthogroups_list", sep = "\t", header = F)
hymnOrthoData <- merge(hymnOrthoData, Pcan.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(hymnOrthoData) <- c("Orthogroup", species[1], species[2], species[3], species[4])
# nrow(hymnOrthoData)

Pdom.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Polistes_dominula_3718_gene_orthogroups_list", sep = "\t", header = F)
hymnOrthoData <- merge(hymnOrthoData, Pdom.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(hymnOrthoData) <- c("Orthogroup", species[1], species[2], species[3], species[4], species[5])
# nrow(hymnOrthoData)

Lfla.ortho <- read.table(file = "~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/orthology-analysis/result/Liostenogaster_flavolineata_3718_gene_orthogroups_list", sep = "\t", header = F)
hymnOrthoData <- merge(hymnOrthoData, Lfla.ortho, by.x = "Orthogroup", by.y = "V1")
colnames(hymnOrthoData) <- c("Orthogroup", species[1], species[2], species[3], species[4], species[5], species[6])
# nrow(hymnOrthoData)
# head(hymnOrthoData)
```

##### Importing & cleaning up the raw count data. Make the associated phenotype metadata.

First, we set a function to do the cleanup to the data & make the associated metadata that we need for DESeq2. For the pheno metadata, make sure that the row names of the metadata dataframe are present and in the same order as the column names of the counts dataframe. Function returns a check to verify that this is true.

```{r}
# df <- infile
# spp <- species[1]
# is.character(spp)
# head(df)
## We are going to change this slightly fx to also filter for the 3718 orthologs

setupDESeq2 <- function(df, spp) {
  
  ## First, filter the df for only those 3718 orthologs:
  df <- df[which(df$Geneid %in% hymnOrthoData[,spp]), ]

  ## Save the gene IDs as a vector for the row names:
  geneID <- df[,1]
  ## Need to remove the geneID column + unecessary second column:
  df <- df[,-c(1:2)]
  ## Rename the headers to the sample names only, stripping extra text:
  names(df) <- gsub("_1Aligned.sortedByCoord.out.bam", "", names(df))
  
  ## Next, move the gene IDs to the row names instead of as a column:
  rownames(df) <- geneID
  
  ## Grab just the columns you want given the metadata:
  cols2grab <- meta[which((meta$R.NR_Phenotype == "R" | meta$R.NR_Phenotype == "NR") & meta$Species == spp),"Sample.Name.on.SRA"]
  ## Select each of the columns from the cols2grab vector:
  df <- df[,as.character(cols2grab)]
  #df %>% select(one_of(cols2grab))

  ## Next, move the gene IDs to the row names instead of as a column:
  rownames(df) <- geneID

  ## Ensure the dimensions look good:
  print(head(df))
  
  ## Now, we also need to make the phenotype metadata. Notice that the resulting pheno DF is analogous to the expression DF
  id <- colnames(df)
  phenotype <- meta[which((meta$R.NR_Phenotype == "R" | meta$R.NR_Phenotype == "NR") & meta$Species == spp),"R.NR_Phenotype"]
  
  pheno <- as.data.frame(phenotype)
  row.names(pheno) <- id

  # Check that the column names of the counts df are the same and in the same order as the phenotypes df:
  print(all(colnames(df) %in% rownames(pheno)))
  print(all(colnames(df) == rownames(pheno)))
  
  return(list(df, pheno)) 
}
```


##### Then, import the raw count data for each species from GitHub and apply the cleanupCounts function:
```{r}
## Import data from GitHub for each species, then process with the cleanupCounts function:

######### Ceratina australensis
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Ceratina-australensis/result/Rehan2018/merged_gene_counts.txt?token=AAVDM4PMRFTRPLQGZT3JWS3ALE5K6"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Ceratina-australensis/result/Rehan2018/merged_gene_counts.txt", header = T, sep = "\t")

Caus <- setupDESeq2(infile, species[1])

######### Ceratina calcarata
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Ceratina-calcarata/result/ShellRehan2019/merged_gene_counts.txt?token=AAVDM4N73PUQBAHSIC7PHKTALE5IQ"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Ceratina-calcarata/result/ShellRehan2019/merged_gene_counts.txt", header = T, sep = "\t")

Ccal <- setupDESeq2(infile, species[2])

######### Megalopta genalis
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Megalopta-genalis/result/merged_gene_counts.txt?token=AAVDM4NGVU62V5CW3C2LLILAKCWDS"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Megalopta-genalis/result/merged_gene_counts.txt", header = T, sep = "\t")

Mgen <- setupDESeq2(infile, species[3])

######### Polistes canadensis
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Polistes-canadensis/result/Patalano/merged_gene_counts.txt?token=AAVDM4LKAXYHR7QA3WQZ7KLAKCODY"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Polistes-canadensis/result/Patalano/merged_gene_counts.txt", header = T, sep = "\t")

Pcan <- setupDESeq2(infile, species[4])

######### Polistes dominula
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Polistes-dominula/result/Taylor/merged_gene_counts.txt?token=AAVDM4MCNBOFO5JB75DXDJ3ALE5NK"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Polistes-dominula/result/Taylor/merged_gene_counts.txt", header = T, sep = "\t")

Pdom <- setupDESeq2(infile, species[5])

######### Liostenogaster flavolineata
# infile <- read.table(text=getURL("https://raw.githubusercontent.com/EmelineFavreau/MajorTransitionScripts/master/comparative-transcriptomics/rnaseq-qc/Liostenogaster-flavolineata/result/Taylor/merged_gene_counts.txt?token=AAVDM4NVAXR26455PI5DBDDAJ5MXS"), header = T, sep = "\t")
infile <- read.table(file="~/Dropbox/GitHub.Local/MajorTransitionScripts/comparative-transcriptomics/rnaseq-qc/Liostenogaster-flavolineata/result/Taylor/merged_gene_counts.txt", header = T, sep = "\t")

Lfla <- setupDESeq2(infile, species[6])

```

##### Quick Summary Statistics on the Raw Feature Counts:
```{r}
countsSummStats <- function(df, spp) {
  header <- c("Total Genes", "Num Genes with >1 read count", "Prop. Genes without Zero Expression", "Num Genes with Zero Expression", "Prop. Genes with Zero Expression", "Prop. Genes with Low Expression", "Num Reads: R Phenotype", "Num Reads: NR Phenotype", "Total Reads Mapped", "N Samples", "Mean CPM per Sample")
  
  ## How many total genes:
  totalGenes <- nrow(df)

  ## One quick quality check: how many of my gene counts are zero for all samples?
  totalCounts <- rowSums(df)
  ## Number of genes with at least 1 count:
  numWithoutZeroExp <- length(totalCounts[totalCounts > 0])
  propWithoutZeroExp <- numWithoutZeroExp / totalGenes
  ## Proportion of all genes not expressed AT ALL:
  numZeroExp <- length(totalCounts[totalCounts == 0])
  propZeroExp <- numZeroExp / totalGenes

  ## How about how many genes where there are fewer than 10 reads in 90% of samples?
  ## If you have 9 samples, and 10 reads per sample, then there should be less than 90 total reads for that gene.
  lowExp <- length(totalCounts[totalCounts < 10*ncol(df)*0.9]) 
  # hist(totalCounts[totalCounts < 10*ncol(df)*0.9])
  # lowExp ## the genes with fewer than 10 counts in 90% of the samples
  ## Now, let's get a proportion of any gene in any sample that was 0:
  propLowExp <- sum(lowExp) / totalGenes

  ## Read count by phenotype
  Rcount <- sum(df[,as.character(meta[which((meta$R.NR_Phenotype == "R") & meta$Species == spp),"Sample.Name.on.SRA"])])
  #Rcount <- sum(df %>% select(one_of(meta[which((meta$R.NR_Phenotype == "R") & meta$Species == spp),"Sample.Name.on.SRA"])))
  NRcount <- sum(df[,as.character(meta[which((meta$R.NR_Phenotype == "NR") & meta$Species == spp),"Sample.Name.on.SRA"])])
  #NRcount <- sum(df %>% select(one_of(meta[which((meta$R.NR_Phenotype == "NR") & meta$Species == spp),"Sample.Name.on.SRA"])))

  # Total read count:
  totalReads <- sum(df)

  # Number of samples
  samples <- ncol(df)
  
  summStats <- as.data.frame(rbind(sprintf("%1.0f", totalGenes), sprintf("%1.0f", numWithoutZeroExp), sprintf("%1.4f", propWithoutZeroExp), sprintf("%1.0f", numZeroExp), sprintf("%1.4f", propZeroExp), sprintf("%1.4f", propLowExp), sprintf("%1.0f", Rcount), sprintf("%1.0f", NRcount), sprintf("%1.0f", totalReads), sprintf("%1.0f", samples), sprintf("%1.2f",totalReads/samples/10^6)))
  rownames(summStats) <- header
  colnames(summStats) <- spp

  return(summStats)
}
df <- Ccal[[1]]
spp <- species[2]
summaryStatsDF <- cbind(countsSummStats(Caus[[1]], species[1]), countsSummStats(Ccal[[1]], species[2]), 
                        countsSummStats(Mgen[[1]], species[3]), countsSummStats(Pcan[[1]], species[4]), 
                        countsSummStats(Pdom[[1]], species[5]), countsSummStats(Lfla[[1]], species[6]))

summaryStatsDF
write.table(summaryStatsDF, file = "orthogroups3718_each/Raw_Read_Counts_Summary_Statistics.txt", row.names = T, col.names = T, append = F, quote = F, sep = "\t")

```

## Resampling N=3 Samples to Generate a More Comparable Estimate of DEG Counts
```{r, echo = FALSE, message= FALSE}
## Set seed
set.seed(50010)

## Make a directory to save everything to:
dir.create(path=paste(getwd(), "/N3_resampling_orthogroups3718", sep=""), showWarnings = TRUE, recursive = FALSE, mode = "0777")

start_time <- Sys.time()

iter = 1000
dgeResampler <- function(hymn, spp) {
  ## Make/reset a df to store it into
  DEG.counts.df <- NULL
  
    ## Make a directory to save everything to:
  dir.create(path=paste(getwd(), "/N3_resampling_orthogroups3718/", spp, "/", sep=""), showWarnings = TRUE, recursive = FALSE, mode = "0777")
  
  ## Subset the data into R and NR
  R <- hymn %>% select(one_of(meta[which((meta$R.NR_Phenotype == "R") & meta$Species == 
                                            spp),"Sample.Name.on.SRA"]))  
  NR <- hymn %>% select(one_of(meta[which((meta$R.NR_Phenotype == "NR") & meta$Species == 
                                            spp),"Sample.Name.on.SRA"]))  
  
  ## Sample 3 times, without replacement from the known columns:
  for (k in 1:iter) {
    tempR <- sample(x = R, size = 3, replace = FALSE) 
    tempNR <- sample(x = NR, size = 3, replace = FALSE)
    hymnResample <- cbind(tempR, tempNR)
    ## Record the raw data that went into each resampling:
    write.table(hymnResample, 
              file=paste("N3_resampling_orthogroups3718/", spp, "/", spp, "Resampled_Raw_Counts_", k, ".txt", sep = ""), 
              sep="\t", 
              quote=F, 
              col.names=NA, 
              row.names = T)

    ## Now make the phenodata object:
    idResample <- as.vector(colnames(hymnResample))
    phenotypeResample <- c(rep("R",3), rep("NR",3))
    phenoResample <- as.data.frame(phenotypeResample)
    row.names(phenoResample) <- idResample
    
    ## Run DESeq2:
    ddsResample <- DESeqDataSetFromMatrix(countData=hymnResample, colData=phenoResample, design=~phenotypeResample)
    ddsResample <- DESeq(ddsResample) ## Yes, overwriting on the dataset obj
    
    ## Extract the normalized counts for this run:
    # normalized_countsResample <- counts(ddsResample, normalized=TRUE)
    # write.table(normalized_countsResample, 
    #             file=paste("N3_resampling_orthogroups3718/resampled_normalized_counts_", k, ".txt", sep = ""),  
    #             sep="\t", 
    #             quote=F, 
    #             col.names=NA, 
    #             row.names = T)
    
    ## Extract the Results table for this run:
    ddsResultsResample <- results(ddsResample, alpha = 0.05)
    write.table(ddsResultsResample, 
                file=paste("N3_resampling_orthogroups3718/", spp, "/", spp, "_Resampled_DESeq_Counts_", k, ".txt", sep = ""), 
                sep="\t", 
                quote=F, 
                col.names=NA, 
                row.names = T)

    ## Now filter to get the number of genes with FDR < 0.05
    sig <- subset(ddsResultsResample, padj < 0.05)
    tempRow <- c(nrow(sig), nrow(subset(sig, log2FoldChange > 0)), nrow(subset(sig, 
                log2FoldChange < 0)), mean(abs(sig$log2FoldChange))) 
    DEG.counts.df <- rbind(DEG.counts.df, tempRow)
  }
  
  ## Make the data frame:
  DEG.counts.df <- as.data.frame(DEG.counts.df)
  rownames(DEG.counts.df) <- c(1:iter)
  colnames(DEG.counts.df) <- c("totalCount", "Up_R", "Up_NR", "mean_abs(LFC)")

  write.table(DEG.counts.df, 
              file = paste("N3_resampling_orthogroups3718/", spp, "/", spp, "_N3_DEG_counts.txt", sep =""), 
              sep="\t", 
              quote = F, 
              col.names = T, 
              row.names = F)
  return(DEG.counts.df)
}

## We will not resample C.calcarata or C. australensis as they are already N = 3; instead we will just need to grab their values.
# Mgen.DEG.counts.df <- dgeResampler(Mgen[[1]], species[3])
# Pcan.DEG.counts.df <- dgeResampler(Pcan[[1]], species[4])
# Pdom.DEG.counts.df <- dgeResampler(Pdom[[1]], species[5])
# Lfla.DEG.counts.df <- dgeResampler(Lfla[[1]], species[6])


end_time <- Sys.time()
paste("RUNTIME:  ", end_time - start_time, " min", sep = "")
```

## Run DESeq2 on the Full dataset for Comparison

#### Run DESeq Function
* **Performs normalization (median of ratios):**
  1.Corrects for variance in read sequencing depth
  2.Corrects for inter-library dispersion in counts (for each gene)

* **Calculates the significance of coefficients with a negative binomial GLM**

Note that DESeq2 does not actually use the normalized counts but rather uses the raw counts and models the normalization in the negative binomial model.

```{r, message = F, warning = F}
runDESeq2 <- function(df, pheno, spp) {
  ## Construct DESEQDataSet Object (stored as dds):
  dds <- DESeqDataSetFromMatrix(countData=df, 
                              colData=pheno, 
                              design=~phenotype)
  ## Design specifies how the counts from each gene depend on our variables in the metadata;    ## we currently only have one factor of interest, phenotype
  ## tidy=TRUE argument tells DESeq2 to output the results table with rownames as a first 
  ## column called 'row; alternatively, I could have used that argument here to move the first   ## col into the rownames of each dataset.
  
  ## Make a directory to save everything to:
  dir.create(path=paste(getwd(), "/orthogroups3718_each/", spp, "/", sep=""), 
             showWarnings = TRUE, recursive = FALSE, mode = "0777")
  
  save(dds, file=paste("orthogroups3718_each/", spp, "/", spp, "_DESeq2_input_matrix.Rdata", 
                       sep=""))

  
  ## Now run the DESeq model:
  dds <- DESeq(dds)
  ## Yes, we are overwriting / performing the functions on the dataset object
  save(dds, file=paste("orthogroups3718_each/", spp, "/", spp, 
                       "_DESeq2_function_output.Rdata", sep=""))

  
  ## Let's extract the normalized counts (from either the DESeq() function or the estimateSizeFactors() function, whichever was used above):
  normalized_counts <- counts(dds, normalized=TRUE)
  write.table(normalized_counts, file=paste("orthogroups3718_each/", spp, "/", spp, 
              "_DESeq2_normalized_counts.txt", sep=""), sep="\t", quote=F, col.names=NA)

  ## I would also like to write the results file for future use / reference:
  ddsResults <- results(dds, contrast = c("phenotype", "R", "NR"))
  write.table(ddsResults, file=paste("orthogroups3718_each/", spp, "/", spp, 
              "_DESeq2_results.txt", sep=""), sep="\t", quote=F, col.names=NA)
  save(ddsResults, file=paste("orthogroups3718_each/", spp, "/", spp, "_DESeq2_results.Rdata",
              sep=""))

  ## Do log-fold change shrinkage as well for volcano plots:
  lfc_shrunk <- lfcShrink(dds, contrast = c("phenotype", "R", "NR"), res = ddsResults, type = 
                          "normal")
  save(lfc_shrunk, file=paste("orthogroups3718_each/", spp, "/", spp, 
                              "_DESeq2_results_lfcShrinkage.Rdata", sep=""))

  ## Phenotype sample sizes
  R.N <- ncol(df[,as.character(meta[which((meta$R.NR_Phenotype == "R") & meta$Species == 
                                            spp),"Sample.Name.on.SRA"])])
  NR.N <- ncol(df[,as.character(meta[which((meta$R.NR_Phenotype == "NR") & meta$Species == 
                                             spp),"Sample.Name.on.SRA"])])
  
  ## Next, I want to filter for the number of genes with FDR < 0.05
  sig <- subset(ddsResults, padj < 0.05)
  DEG.counts.df <- c(sprintf("%1.0f", nrow(sig)), sprintf("%1.0f", nrow(subset(sig, 
                    log2FoldChange > 0))), sprintf("%1.0f", nrow(subset(sig, log2FoldChange 
                    < 0))), sprintf("%1.3f", mean(abs(sig$log2FoldChange))), sprintf("%1.3f", 
                    sd(abs(sig$log2FoldChange))), R.N, NR.N) 
  DEG.counts.df <- as.data.frame(DEG.counts.df)
  rownames(DEG.counts.df) <- c("Total DEGs", "Up-Reg in R", "Up-Reg in NR", "meanAbsLFC", 
                               "stdevAbsLFC", "Reproductives N", "Non-Repros N")
  colnames(DEG.counts.df) <- spp

  ## Lastly, we want the identities of the DEGs:
  allDEGs <- rownames(sig)
  write.table(allDEGs, file = paste("orthogroups3718_each/", spp, "/", spp, "_all_DEGs.txt", 
                                    sep=""), sep="\t", quote=F, col.names=F, row.names = F)
  upR <- rownames(subset(sig, log2FoldChange > 0))
  write.table(upR, file = paste("orthogroups3718_each/", spp, "/", spp, "_upR_DEGs.txt", 
                                sep=""), sep="\t", quote=F, col.names=F, row.names = F)
  upNR <- rownames(subset(sig, log2FoldChange < 0))
  write.table(upNR, file = paste("orthogroups3718_each/", spp, "/", spp, "_upNR_DEGs.txt", 
                                 sep=""), sep="\t", quote=F, col.names=F, row.names = F)

  return(list(dds, normalized_counts, ddsResults, DEG.counts.df, allDEGs, upR, upNR))
}

## Note that size factors for normalization could be extracted as follows:
#dds <- estimateSizeFactors(dds)
#sizeFactors(dds)
# spp = species[2]
# df = Ccal[[1]]
# pheno = Ccal[[2]]
Caus_DEG <- runDESeq2(Caus[[1]], Caus[[2]], species[1])
Ccal_DEG <- runDESeq2(Ccal[[1]], Ccal[[2]], species[2])
Mgen_DEG <- runDESeq2(Mgen[[1]], Mgen[[2]], species[3])
Pcan_DEG <- runDESeq2(Pcan[[1]], Pcan[[2]], species[4])
Pdom_DEG <- runDESeq2(Pdom[[1]], Pdom[[2]], species[5])
Lfla_DEG <- runDESeq2(Lfla[[1]], Lfla[[2]], species[6])

## Let's also make a vector of the dds objects so we can loop through them downstream:
dds.list <- c(Caus_DEG[1], Ccal_DEG[1], Mgen_DEG[1], Pcan_DEG[1], Pdom_DEG[1], Lfla_DEG[1])

combinedDEGcounts <- cbind(Caus_DEG[[4]], Ccal_DEG[[4]], Mgen_DEG[[4]], Pcan_DEG[[4]], 
                           Pdom_DEG[[4]], Lfla_DEG[[4]])
combinedDEGcounts

write.table(combinedDEGcounts, file = "orthogroups3718_each/DEGs_Summary_All_Species.txt", 
            row.names = T, col.names = T, append = F, quote = F, sep = "\t")

```

#### Run Variance-Stabilizing Transformation
We will also need to run the VST for downstream visualization and applications like WGCNA, which will include filtering out low expression values too (prints files with and without filtering). We are using a parametric fit with the VST. **Note that all transformations have set blind = FALSE. We do not want the transformations to be blind to the experimental design (phenotypes) at this point; we expect large differential expression in some genes, and so we need to control for outliers by setting blind to FALSE.**

A note on the filtering: Langfelder has said: "I require that a gene has a relatively high expression (e.g., 0.5 to 1 count per million reads, this translates to a counts in low tens for a typical data set with 30-50M reads per sample) in at least 1/4 of the samples (or whatever fraction is the smallest experimental group of the design). The rationale is that typical correlation analysis in WGCNA assumes (approximately) continuous data; using correlation on counts below say 5-10 which tend to be mostly zero can really lead to spurious results."

So, to apply the same logic here, we have 7-15 CPM per sample, which is less than half of the read depth that Langfelder typically has. So, we will filter for at least 5 in 1/2 of the samples. 

```{r}
runVST <- function(dds, samples, spp) {
    
    ## Run the VST without filtering (for SVM):
    vsdata1 <- varianceStabilizingTransformation(dds, blind = FALSE, fitType = "parametric")
    jpeg(filename = paste("orthogroups3718_each/", spp, "/", spp, "_VST_All_Genes.jpg", sep=""), 
         width = 8, height = 8, units = "in", res = 400)
    meanSdPlot(assay(vsdata1))
    dev.off()
    write.table(assay(vsdata1), file=paste("orthogroups3718_each/", spp, "/", spp, "_VST.txt",
                                           sep=""), sep="\t", quote=F, row.names=T)
    save(vsdata1, file=paste("orthogroups3718_each/", spp, "/", spp, "_VST.Rdata", sep=""))

    ## Run the VST with filtering (for WGCNA):
    keep <- rowSums(counts(dds) >= 5) >= samples/2    ## This filters for 5 in the smallest 
                                                      ## group size
    print(paste("I filtered ", nrow(dds) - nrow(dds[keep,]), " genes from the dataset.\n", 
                sep=""))
    dds.filt <- dds[keep,]
    vsdata2 <- varianceStabilizingTransformation(dds.filt, blind = FALSE, fitType = 
                                                   "parametric")
    jpeg(filename = paste("orthogroups3718_each/", spp, "/", spp, 
                          "_VST_After_Filtering_Low_Expr.jpg", 
                          sep=""), width = 8, height = 8, units = "in", res = 400)
    meanSdPlot(assay(vsdata2))
    dev.off()
    write.table(assay(vsdata2), file=paste("orthogroups3718_each/", spp, "/", spp, 
                                           "_VST_lowExprfiltered.txt", sep=""), sep="\t",
                                            quote=F, row.names=T)
    name <- paste(spp, ".vsdata", sep="")
    save(vsdata2, file=paste("orthogroups3718_each/", spp, "/", spp, 
                             "_VST_lowExprfiltered.Rdata", sep=""))
  }
# dds <- dds.list[[1]]
# samples <- ncol(Caus[[1]])
# spp <- species[2]

Caus.vst <- runVST(dds.list[[1]], ncol(Caus[[1]]), species[1])
Ccal.vst <- runVST(dds.list[[2]], ncol(Ccal[[1]]), species[2])
Mgen.vst <- runVST(dds.list[[3]], ncol(Mgen[[1]]), species[3])
Pcan.vst <- runVST(dds.list[[4]], ncol(Pcan[[1]]), species[4])
Pdom.vst <- runVST(dds.list[[5]], ncol(Pdom[[1]]), species[5])
Lfla.vst <- runVST(dds.list[[6]], ncol(Lfla[[1]]), species[6])
```
