---
title: "TopGo analysis of DEG using PANTER GO Slim Terms BP"
author: "Emeline Favreau"
date: "06/05/2021"
output: html_document
---
Copyright 2021 Emeline Favreau, University College London.


###### Objective of analysis: Among the 3 wasp and 3 bee species, are there common pattern of GO Slim term enrichment?

We obtained PANTER GO Slim Terms BP that are overprepresented in genes subsets. Specifically, we focused on genes that are differentially expressed between reproductives and non-reproductives in _Ceratina australensis_, _C. calcarata_, _Megalopta genalis_, _Polistes canadensis_, _P. dominula_ and _Liostenogaster flavolineata_.


###### Analysis steps:
- Obtaining data
- Aim 1: Biological processes enrichment in all DEG





```{r load all the libraries, eval = TRUE, echo = FALSE, include = FALSE}
# get libraries
basic_libraries <- c("ggplot2",
                     "tidyverse",
                     "readr")

for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```

```{r common groups}


# the groups that we are testing
all_species <- c("Polistes_canadensis",
          "Polistes_dominula",
          "Liostenogaster_flavolineata",
          "Ceratina_australensis",
          "Ceratina_calcarata" ,
          "Megalopta_genalis")

species_present <- c("Ceratina_australensis",
          "Ceratina_calcarata" ,
          "Liostenogaster_flavolineata")



```

##### Aim 1: Biological processes enrichment in all significant DEG


```{r import data BP all sig DEG, eval = TRUE, echo = FALSE, include = FALSE}
# GO terms enrichment result tables from PANTHER overpresentation test
# file obtained by PANTHER GO Slim Terms (BP) overrepresentation test, containing 8 columns
# GOTerm REFLIST DEGs	DEGs_expected DEGs_over_under DEGs_fold_Enrichment DEGs_raw_P_value DEGs_FDR
# these are Biological processes for DEG upregulated in Reproductives FDR < 0.05
Ceratina_australensis_BP <- read_delim("resultOnScratch/DEGs-Ceratina_australensis_PANTHER-GOSlim-BP-overrepresentation-test-FDR-results-trimmed.txt", "\t", escape_double = FALSE,
                                    col_names = FALSE, trim_ws = TRUE)

Ceratina_calcarata_BP <- read_delim("resultOnScratch/DEGs-Ceratina_calcarata_PANTHER-GOSlim-BP-overrepresentation-test-FDR-results-trimmed.txt", "\t", escape_double = FALSE,
                                    col_names = FALSE, trim_ws = TRUE)


Liostenogaster_flavolineata_BP <- read_delim("resultOnScratch/DEGs-Liostenogaster_flavolineata_PANTHER-GOSlim-BP-overrepresentation-test-FDR-results-trimmed.txt", "\t", escape_double = FALSE,
                                    col_names = FALSE, trim_ws = TRUE)

```

```{r make table BP , eval = TRUE, echo = FALSE, include = TRUE}
# name columns
colnames(Ceratina_australensis_BP) <- c("GOTerm", "REFLIST", "DEGs",
                                        "DEGs_expected", "DEGs_over_under",
                                        "DEGs_fold_Enrichment",
                                        "DEGs_raw_P_value", "DEGs_FDR")

colnames(Ceratina_calcarata_BP) <- c("GOTerm", "REFLIST", "DEGs",
                                        "DEGs_expected", "DEGs_over_under",
                                        "DEGs_fold_Enrichment",
                                        "DEGs_raw_P_value", "DEGs_FDR")

colnames(Liostenogaster_flavolineata_BP) <- c("GOTerm", "REFLIST", "DEGs",
                                        "DEGs_expected", "DEGs_over_under",
                                        "DEGs_fold_Enrichment",
                                        "DEGs_raw_P_value", "DEGs_FDR")

# make a colum for GO.ID
Ceratina_australensis_BP$GO.ID <- gsub(x = Ceratina_australensis_BP$GOTerm,
                                       pattern = ".*\\(|\\)",
                                       replacement = "")

Ceratina_calcarata_BP$GO.ID <- gsub(x = Ceratina_calcarata_BP$GOTerm,
                                       pattern = ".*\\(|\\)",
                                       replacement = "")

Liostenogaster_flavolineata_BP$GO.ID <- gsub(x = Liostenogaster_flavolineata_BP$GOTerm,
                                       pattern = ".*\\(|\\)",
                                       replacement = "")


# make a colum for species
Ceratina_australensis_BP$species <- "Ceratina_australensis"
Ceratina_calcarata_BP$species <- "Ceratina_calcarata"
Liostenogaster_flavolineata_BP$species <- "Liostenogaster_flavolineata"



# result table: number of over/underpresented GO SlimTerms for each species
result_BP_df  <-  as.data.frame(rbind(c("Ceratina_australensis",
                                        nrow(Ceratina_australensis_BP)),
                                      c("Ceratina_calcarata",
                                        nrow(Ceratina_calcarata_BP)),
                                      c("Liostenogaster_flavolineata",
                                        nrow(Liostenogaster_flavolineata_BP))),
    stringsAsFactors = FALSE)

# name coluns
colnames(result_BP_df) <- c("species", "over_under_represented_SlimGO_Term_in_BP")

# print
result_BP_df

# obtain vector of all GO
GO.ID_BP <- unique(c(Ceratina_australensis_BP$GO.ID,
                     Ceratina_calcarata_BP$GO.ID,
                     Liostenogaster_flavolineata_BP$GO.ID))


# make a new table
# columns: GOID, Terms, species
stacked_result_BP <- rbind(Ceratina_australensis_BP,
                     Ceratina_calcarata_BP,
                     Liostenogaster_flavolineata_BP)

# obtain frequencies of GO Terms for all aspecies
test <- stacked_result_BP

# dataframe
# GOID | Species 1 | Species 2 | Species 3 | 
wide_df <- as.data.frame(matrix(NA, nrow = length(unique(test$GO.ID)),
                                ncol = 4))
# name the columns
colnames(wide_df) <- c("GO.ID", species_present)

# make a vector with all GO terms
go_vec <- unique(test$GO.ID)

# assign thios go terms to the column
wide_df$GO.ID <- go_vec

# set all species column count to 0
wide_df$Liostenogaster_flavolineata <- 0
wide_df$Ceratina_australensis       <- 0
wide_df$Ceratina_calcarata          <- 0


for(i in go_vec){
  
  # subset data for a given GO Term ID
  this_subset <- test %>% filter(GO.ID == i)
  
  # assign 1 if present in this species
  if("Ceratina_australensis" %in% this_subset$species){
    wide_df$Ceratina_australensis[wide_df$GO.ID == i] <- 1
  }
  
  if("Polistes_canadensis" %in% this_subset$species){
    wide_df$Polistes_canadensis[wide_df$GO.ID == i] <- 1
  }
  
  if("Polistes_dominula" %in% this_subset$species){
    wide_df$Polistes_dominula[wide_df$GO.ID == i] <- 1
  }
  
  if("Liostenogaster_flavolineata" %in% this_subset$species){
    wide_df$Liostenogaster_flavolineata[wide_df$GO.ID == i] <- 1
  }
  
  if("Ceratina_calcarata" %in% this_subset$species){
    wide_df$Ceratina_calcarata[wide_df$GO.ID == i] <- 1
  }
  
  if("Megalopta_genalis" %in% this_subset$species){
    wide_df$Megalopta_genalis[wide_df$GO.ID == i] <- 1
  }
}


# add a column with count of species that have that GO term  
wide_df$species_count <- wide_df %>%
  select(all_of(species_present)) %>%
  rowSums()

# add a column with Term
wide_df$GOTerm <- test$GOTerm[match(wide_df$GO.ID, test$GO.ID)]

# add a column for BP
wide_df$GOCategory <- "PANTHER-GO-Slim-BP"

# save for supplementary info
write.table(x         = wide_df,
            file      = "resultOnScratch/DEGs_PANTHER_BP_GOSlimTerms.csv",
            quote     = FALSE,
            row.names = FALSE,
            sep       = ",")
```

11 overlap between species in GO Slim Terms.

```{r exploration of BP results}
## exploration of BP results
# 11 overlap between species in GO Slim Terms.
# unique(str(wide_df$species_count))
wide_df %>% filter(species_count == 2) %>%
  filter(Ceratina_australensis == 1) %>%
  filter(Liostenogaster_flavolineata == 1) %>% 
  select(GOTerm)

# all under-represented in DEGs of australensis and flavolineata
#macromolecule biosynthetic process (GO:0009059)			
#cellular macromolecule biosynthetic process (GO:0034645)			
#regulation of metabolic process (GO:0019222)			
#regulation of nitrogen compound metabolic process (GO:0051171)			
#regulation of primary metabolic process (GO:0080090)			
#regulation of macromolecule metabolic process (GO:0060255)			
#cellular macromolecule metabolic process (GO:0044260)			
#gene expression (GO:0010467)			
#macromolecule metabolic process (GO:0043170)			
#RNA metabolic process (GO:0016070)
```




**Conclusion**



```{r record versions of session, eval = TRUE, echo = FALSE, include = FALSE}
# record versions of R and packages here
sessionInfo()
# R version 3.6.3 (2020-02-29)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Catalina 10.15.4
# 
# Matrix products: default
# BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] forcats_0.5.0   stringr_1.4.0   dplyr_0.8.5     purrr_0.3.4     readr_1.3.1    
# [6] tidyr_1.0.3     tibble_3.0.1    tidyverse_1.3.0 ggplot2_3.3.0  
# loaded via a namespace (and not attached):
#  [1] Rcpp_1.0.4.6     cellranger_1.1.0 pillar_1.4.4     compiler_3.6.3   dbplyr_1.4.3    
#  [6] tools_3.6.3      lubridate_1.7.8  jsonlite_1.6.1   lifecycle_0.2.0  gtable_0.3.0    
# [11] nlme_3.1-147     lattice_0.20-41  pkgconfig_2.0.3  rlang_0.4.6      reprex_0.3.0    
# [16] cli_2.0.2        DBI_1.1.0        rstudioapi_0.11  yaml_2.2.1       haven_2.2.0     
# [21] xfun_0.13        xml2_1.3.2       withr_2.2.0      httr_1.4.1       knitr_1.28      
# [26] fs_1.4.1         generics_0.0.2   vctrs_0.3.0      hms_0.5.3        grid_3.6.3      
# [31] tidyselect_1.1.0 glue_1.4.1       R6_2.4.1         fansi_0.4.1      readxl_1.3.1    
# [36] modelr_0.1.7     magrittr_1.5     scales_1.1.1     backports_1.1.7  ellipsis_0.3.0  
# [41] rvest_0.3.5      assertthat_0.2.1 colorspace_1.4-1 stringi_1.4.6    munsell_0.5.0   
# [46] broom_0.5.6      crayon_1.3.4 
```
