---
title: "Orthogroup Hash Tables"
author: "EF"
date: "3/16/2021"
output: html_document
---

Just checking if 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
## Clear workspace
rm(list=ls())

## Set options
options(scipen=999)

## Load libraries
library(dplyr)
library(tidyverse)
# BiocManager::install("rtracklayer")
library(rtracklayer)  ## Need to read in GFF
```

## Read in the Bees Only orthogroup Info:
```{r}

# load list of orthologs
bees1to1 <- read.table(file = "../result/BeesDrosoApis/Orthogroups/Orthogroups_SingleCopyOrthologues.txt")
head(bees1to1)
dim(bees1to1)
# checking classes, important when merging
class(bees1to1$V1)

# load table of orthogroups
beesOrthogroups <- read.table(file = "../result/BeesDrosoApis/Orthogroups/Orthogroups.txt",
                              sep = ":")
head(beesOrthogroups)
dim(beesOrthogroups)
str(beesOrthogroups)


## Now, I know that the orthologs are all listed in one column; before we deal with that, let's merge with the 1:1's:
bees1to1merged <- merge(bees1to1,
                        beesOrthogroups,
                        by = "V1",
                        all.x = T)
dim(bees1to1merged)
head(bees1to1merged)

# grep(x = beesOrthogroups,
#      pattern = bees1to1merged$V1,
#      value = TRUE)
## There is trailing whitespace that is being made into the first column
## Species were associated to the GFF for gene ID
bees1to1merged <- bees1to1merged %>% separate(V2, c("blank",
                                                    "C_australensis",
                                                    "D_melanogaster",
                                                    "A_mellifera",
                                                    "C_calcarata",
                                                    "M_genalis"),
                                              sep = " ")
beesOrthologs <- bees1to1merged[,-2]
colnames(beesOrthologs)[1] <- "Orthogroup"
head(beesOrthologs)
str(beesOrthologs)
beesOrthologs$Orthogroup <- as.character(beesOrthologs$Orthogroup)
```

## Some of the bee species' gene IDs in the transcriptomics data are different IDs from the GFF file, so we need to associate them. We do not need to do C. australensis, though.

###### C. calcarata -- associate locus ID with the ID in orthogroup
```{r}
## I only want to load some of the tags from the GFF:
my_tags <- c("ID", "Parent", "Name", "Dbxref", "geneID", "gene")

Ccal.gff <- readGFF("../../rnaseq-qc/Ceratina-calcarata/input/GCF_001652005.1_ASM165200v1_genomic.gff.gz", tags = my_tags)
# head(Ccal.gff)
# names(Ccal.gff)
Ccal.gff.filt <- as.data.frame(Ccal.gff[which(Ccal.gff$type == "CDS"),
                                        c("gene", "Name")])
# head(Ccal.gff.filt)
## At this point, we have non-unique entries, so we need to condense this:
Ccal.gff.filt.uniq <- distinct(Ccal.gff.filt, Name, gene) 

beesOrthologs_merged <- merge(beesOrthologs,
                       Ccal.gff.filt.uniq,
                       by.x = "C_calcarata",
                       by.y = "Name",
                       all.x = T)


#names(beesOrthologs_merged)
colnames(beesOrthologs_merged)[7] <- "C_calcarata_locusID"

# check result: 323 obs. of  7 variables (one column more than before)
str(beesOrthologs_merged)
str(beesOrthologs)


```