---
title: "Six species tree"
author: "Emeline Favreau"
date: "2021/03/23"
output: html_document
---

Copyright 2021 Emeline Favreau, University College London.

---
# Objective of analysis

Make a simple tree for Cold Spring Harbor Meeting poster

## Analysis steps:
- Obtaining data
- Aim 1: description
- Aim 2: description
- Aim 3: description
- Aim 4: description



```{r load all the libraries, eval = TRUE, echo = FALSE, include = FALSE}
# get libraries
basic_libraries <- c("ggplot2",
                     "tidyverse",
                     "readxl",
                     "phytools")

for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}

#BiocManager::install("treeio")
library("treeio")
library("ggtree")
```


```{r import data, eval = TRUE, echo = FALSE, include = FALSE}
# tree file, newick format
# made by OrthoFinder using single orthologous data
ortho_tree <- read.newick(file = "result/allSpecies/tree/SpeciesTree_rooted.txt")
```



```{r aim 1, eval = TRUE, echo = FALSE, include = TRUE}

# data to drop 
to_drop <- c("Drosophila_melanogaster",
             "Apis_mellifera-longest-isoforms")

# display tree only with species I want
ortho_tree_reduced <- drop.tip(ortho_tree,
                         to_drop)

# # make a tree
# ortho_tree_reduced %>% 
#   ggtree() + 
#   geom_tiplab() +
#   theme_tree2() +
#   ggtitle("OrthoFinder Tree")

# tidy ttp label
ortho_tree_reduced$tip.label <- gsub(pattern = "-longest-isoforms",
     x = ortho_tree_reduced$tip.label,
     replacement = "")

ortho_tree_reduced$tip.label <- gsub(pattern = "_",
     x = ortho_tree_reduced$tip.label,
     replacement = " ")


# make a list for each family, in each element, the names of the tips
taxonomy_list <- list("Apidae"         = c("Ceratina australensis","Ceratina calcarata"),
                      "Halictidae"     = "Megalopta genalis",
                      "Vespidae"       = c("Polistes dominula", "Polistes canadensis"),
                      "Stenogastrinae" = "Liostenogaster flavolineata")

# update tree with group information
ortho_tree_reduced_group_info <- groupOTU(ortho_tree_reduced,
                  taxonomy_list)
# colour pick
#bee orange #f1a340 #e34a33
#wasps purple #998ec3 #bcbddc
# plot tree with coloured branches
ggtree(ortho_tree_reduced_group_info, aes(color = group)) +
  geom_tiplab(size = 3) +
  geom_treescale(x = 0, fontsize = 3) +
  scale_color_manual(values = c("#f1a340", "#e34a33", "#fdae6b", "#756bb1", "#bcbddc")) +
  #geom_point(aes(shape=isTip, color=isTip), size=3) +
  ggtitle("Single-orthologous genes tree",
          subtitle = "from OrthoFinder") +
  theme(legend.position = "none") + xlim(NA, 0.4)
```



```{r aim 2, eval = TRUE, echo = FALSE, include = TRUE}
# plot tree with dots on nodes
ggtree(ortho_tree_reduced_group_info) +
  geom_tiplab(size = 3) +
  geom_treescale(x = 0, fontsize = 3) +
  geom_nodepoint(size = 3) +
  ggtitle("Single-orthologous genes tree",
          subtitle = "from OrthoFinder") +
  theme(legend.position = "none") + xlim(NA, 0.4)
```

Conclusion

## Aim 3: description

```{r aim 3, eval = TRUE, echo = FALSE, include = TRUE}
# plot tree for figure
ggtree(ortho_tree_reduced_group_info) +
  geom_tiplab(size = 3) +
  geom_treescale(x = 0, fontsize = 3) +
  theme(legend.position = "none") + xlim(NA, 0.4)

ggtree(ortho_tree_reduced_group_info, branch.length="none") +
  geom_tiplab(size = 6) + xlim(NA, 7)

ggsave(filename = "result/2021-03-23-cladogram-from-orthofinder.jpg",
       height = 70,
       width = 170,
       units = "mm")
```

Conclusion

## Aim 4: description

```{r aim 4, eval = TRUE, echo = FALSE, include = TRUE}
# perform analysis 
# make a figure 
```

Conclusion

```{r record versions of session, eval = TRUE, echo = FALSE, include = FALSE}
# record versions of R and packages here
sessionInfo()
# R version 3.6.3 (2020-02-29)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Catalina 10.15.4
# 
# Matrix products: default
# BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
# 
# locale:
# [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] forcats_0.5.0   stringr_1.4.0   dplyr_0.8.5     purrr_0.3.4     readr_1.3.1    
# [6] tidyr_1.0.3     tibble_3.0.1    tidyverse_1.3.0 ggplot2_3.3.0  
# loaded via a namespace (and not attached):
#  [1] Rcpp_1.0.4.6     cellranger_1.1.0 pillar_1.4.4     compiler_3.6.3   dbplyr_1.4.3    
#  [6] tools_3.6.3      lubridate_1.7.8  jsonlite_1.6.1   lifecycle_0.2.0  gtable_0.3.0    
# [11] nlme_3.1-147     lattice_0.20-41  pkgconfig_2.0.3  rlang_0.4.6      reprex_0.3.0    
# [16] cli_2.0.2        DBI_1.1.0        rstudioapi_0.11  yaml_2.2.1       haven_2.2.0     
# [21] xfun_0.13        xml2_1.3.2       withr_2.2.0      httr_1.4.1       knitr_1.28      
# [26] fs_1.4.1         generics_0.0.2   vctrs_0.3.0      hms_0.5.3        grid_3.6.3      
# [31] tidyselect_1.1.0 glue_1.4.1       R6_2.4.1         fansi_0.4.1      readxl_1.3.1    
# [36] modelr_0.1.7     magrittr_1.5     scales_1.1.1     backports_1.1.7  ellipsis_0.3.0  
# [41] rvest_0.3.5      assertthat_0.2.1 colorspace_1.4-1 stringi_1.4.6    munsell_0.5.0   
# [46] broom_0.5.6      crayon_1.3.4 
```
